unified-scheduler-logic/src/lib.rs

```diff
@@ -1,20 +1,1299 @@
+        self.unblocked_task_queue.pop_front().inspect(|_| {
+            self.unblocked_task_count.increment_self();
+        })
+    }
+
+    pub fn deschedule_task(&mut self, task: &Task) {
+        self.active_task_count.decrement_self();
+        self.handled_task_count.increment_self();
+        self.unlock_for_task(task);
+    }
+
+    fn try_lock_usage_queue(
+        usage_queue: &UsageQueueInner,
+        requested_usage: RequestedUsage,
+    ) -> LockResult {
+        match usage_queue.current_usage {
+            Usage::Unused => LockResult::Ok(Usage::from(requested_usage)),
+            Usage::Readonly(count) => match requested_usage {
+                RequestedUsage::Readonly => LockResult::Ok(Usage::Readonly(count.increment())),
+                RequestedUsage::Writable => LockResult::Err(()),
+            },
+            Usage::Writable => LockResult::Err(()),
+        }
+    }
+
+    #[must_use]
+    fn unlock_usage_queue(
+        usage_queue: &mut UsageQueueInner,
+        attempt: &LockAttempt,
+    ) -> Option<(RequestedUsage, Task)> {
+        let mut is_unused_now = false;
+        match &mut usage_queue.current_usage {
+            Usage::Readonly(ref mut count) => match attempt.requested_usage {
+                RequestedUsage::Readonly => {
+                    if count.is_one() {
+                        is_unused_now = true;
+                    } else {
+                        count.decrement_self();
+                    }
+                }
+                RequestedUsage::Writable => unreachable!(),
+            },
+            Usage::Writable => match attempt.requested_usage {
+                RequestedUsage::Writable => {
+                    is_unused_now = true;
+                }
+                RequestedUsage::Readonly => unreachable!(),
+            },
+            Usage::Unused => unreachable!(),
+        }
+
+        if is_unused_now {
+            usage_queue.current_usage = Usage::Unused;
+            usage_queue.pop_unblocked_usage_from_task()
+        } else {
+            None
+        }
+    }
+
+    #[must_use]
+    fn try_lock_for_task(&mut self, task: Task) -> Option<Task> {
+        let mut blocked_usage_count = ShortCounter::zero();
+
+        for attempt in task.lock_attempts() {
+            let usage_queue = attempt.usage_queue_mut(&mut self.usage_queue_token);
+            let lock_result = if usage_queue.has_no_blocked_usage() {
+                Self::try_lock_usage_queue(usage_queue, attempt.requested_usage)
+            } else {
+                LockResult::Err(())
+            };
+            match lock_result {
+                LockResult::Ok(Usage::Unused) => unreachable!(),
+                LockResult::Ok(new_usage) => {
+                    usage_queue.current_usage = new_usage;
+                }
+                LockResult::Err(()) => {
+                    blocked_usage_count.increment_self();
+                    let usage_from_task = (attempt.requested_usage, task.clone());
```

Comment:
oh, I like `.clone()` being actually `Arc::clone()` behind scene to hide this (minor) implementation detail.

Reply:
Agree to disagree on this one I guess - when I see a clone() I think "this might be expensive and there might be an optimization possible in removing the clone", while if I see Arc::clone() I know it's not expensive.

Reply:
> if I see Arc::clone() I know it's not expensive.

hehe, fun fact is that `Arc::clone()` is super slow in this unified scheduler code.

Reply:
"super slow" is a bold claim that requires extraordinary evidence 😉

Reply:
hehe, true.

how about this? i think 1.7x perf increase should be a big deal: https://github.com/solana-labs/solana/pull/35286#issuecomment-1964383019 just with this code changes https://github.com/ryoqun/solana/pull/17#discussion_r1536575150 (with `unsafe`, unfortunately)